stdlib/asm/collections/smt.masm

```diff
@@ -1,14 +1,572 @@
+#!
+#! Returns the node value (hash) of the previous leaf at the given index, as well as
+#! the root for the new SMT.
+#!
+#! Inputs: [start_addr, end_addr, leaf_index, R]
+#! Outputs: [NV_old, R_new]
+proc.hash_and_insert_mem
+    # => [start_addr, end_addr, i, R, ...]
+
+    dup.1 dup.1
+    # => [start_addr, end_addr, start_addr, end_addr, i, R, ...]
+
+    exec.rpo::hash_memory_words
+    # => [NV, start_addr, end_addr, i, R, ...]
+
+    # We're already ready for adv.insert_mem.
+    adv.insert_mem
+
+    # Prepare for mtree_set
+    movup.4 drop movup.4 drop
+    # => [NV, i, R, ...]
+
+    movup.4 movdn.8
+    # => [NV, R, i]
+
+    swapw movup.8
+    # => [i, R, NV]
+
+    push.LEAF_DEPTH
+    # => [d, i, R, NV]
+
+    mtree_set
+    # => [NV_old, R_new, ...]
+end
+
+#! Inserts a key-value pair into a leaf currently containing a single, non-empty pair,
+#! converting the leaf to a multi-leaf.
+#!
+#! Inputs: [K_in_leaf, K, V_in_leaf, V, R]
+#! Outputs: [V_in_leaf, R_new]
+proc.insert_single_to_multi_leaf.16
+
+    # Pairs in leaves are sorted by key, so we need to compare the new key against the existing
+    # key.
+
+    dupw dupw.2
+    # => [K, K_l, K_l, K, V_l, V, R]
+
+    exec.word::lt
+    # => [K_l < K, K_l, K, V_l, V, R]
+
+    # We need a contiguous region of memory containing the sorted pairs. Converting a single-leaf
+    # To a multi-leaf will always involve 2 pairs (8 words), so we can use local memory for this.
+
+    if.true
+        # => [K_l, K, V_l, V, R]
+
+        # Place our locals as follows:
+        #   0..4: K_in_leaf
+        #   4..8: V_in_leaf
+        #  8..12: K
+        # 12..16: V
+
+        loc_storew.0 dropw
+        # => [K, V_l, V, R]
+
+        # Save K[3], as that is our index `i`.
+        dup movdn.12
+        # => [K, V_l, V, i, R]
+
+        loc_storew.8 dropw
+        # => [V_l, V, i, R]
+
+        loc_storew.4 dropw
+        # => [V, i, R]
+
+        loc_storew.12 dropw
+        # => [i, R]
+    else
+        # => [K_l, K, V_l, V, R]
+
+        # Place our locals as follows:
+        #   0..4: K
+        #   4..8: V
+        #  8..12: K_in_leaf
+        # 12..16: V_in_leaf
+
+        loc_storew.8 dropw
+        # => [K, V_l, V, R]
+
+        # Save K[3], as that is our index `i`.
+        dup movdn.12
+        # => [K, V_l, V, i, R]
+
+        loc_storew.0 dropw
+        # => [V_l, V, i, R]
+
+        loc_storew.12 dropw
+        # => [V, i, R]
+
+        loc_storew.4 dropw
+        # => [i, R]
+    end
+
+    # => [i, R]
+
+    # Prepare for hash_memory_words,
+    locaddr.12 add.4
+    locaddr.0
+    # => [start_addr, end_addr, i, R]
+
+    exec.hash_and_insert_mem
+    # => [NV_old, R_new]
+
+    # By nature, if an insertion transforms a single-leaf to a multi-leaf, then the old value at
+    # this key is an empty word.
+
+    padw swapw dropw
+    # => [Z, R_new]
+end
+
+
+# MAX_LEAF_SIZE + 4
+proc.insert_into_leaf.8196
```

Comment:
Noticed the O(n) memory copying pattern in `insert_into_leaf` and `remove_in_multi_leaf`. Curious about your thoughts on when we might need to optimize for leaves with many pairs - thinking ahead now could help avoid breaking changes later if we need a hybrid approach with advice provider. 

Not asking for any changes to this PR, just wondering if it makes sense to have e.g. caps in input sizes here.

Reply:
> I think we do test against the limit 1024 in MASM

Can you expand? Do you mean somewhere in this `smt.masm` file, we run that check (in which case I missed it on my first read)?